Source/WebCore/platform/text/TextSpacing.cpp

```diff
@@ -0,0 +1,110 @@
+        ++result;
+    return result;
+}
+
+
+namespace TextSpacing {
+
+static bool isIdeograph(char32_t character)
+{
+    // All characters in the range of U+3041 to U+30FF, except those that belong to Unicode Punctuation [P*]
+    if ((character >= 0x3041 && character <= 0x30FF) && !isPunctuation(character))
+        return true;
+    // CJK Strokes (U+31C0 to U+31EF).
+    if (character >= 0x31C0 && character <= 0x31EF)
+        return true;
+    // Katakana Phonetic Extensions (U+31F0 to U+31FF)
+    if (character >= 0x31F0 && character <= 0x31FF)
+        return true;
+    if (isOfScriptType(character, UScriptCode::USCRIPT_HAN))
+        return true;
+
+    return false;
+}
+
+static bool isFullWidthOpeningPunctuation(char32_t character, uint32_t generalCategoryMask)
+{
+    if (isOpeningPunctuation(generalCategoryMask) && (isCJKSymbolOrPunctuation(character) || isEastAsianFullWidth(character)))
+        return true;
+    return character == leftSingleQuotationMark || character == leftDoubleQuotationMark;
+}
+
+static bool isFullWidthClosingPunctuation(char32_t character, uint32_t generalCategoryMask)
+{
+    if (isClosingPunctuation(generalCategoryMask) && (isCJKSymbolOrPunctuation(character) || isEastAsianFullWidth(character)))
+        return true;
+    return character == rightSingleQuotationMark || character == rightDoubleQuotationMark;
+}
+
+static bool isNonIdeographicNumeral(char32_t character, uint32_t generalCategoryMask)
+{
+    // FIXME: Should also check that it is not: upright in vertical text flow using the text-orientation property or the text-combine-upright property.
+    return ((generalCategoryMask & U_GC_ND_MASK) && !isEastAsianFullWidth(character));
+}
+
+// Classes are defined at https://www.w3.org/TR/css-text-4/#text-spacing-classes
+CharacterClass characterClass(char32_t character)
+{
+    auto generalCategoryMask = U_GET_GC_MASK(character);
+    if (isIdeograph(character))
+        return CharacterClass::Ideograph;
+
+    // We already know it is not an Ideograph from here
+    // FIXME: Should also check that it is not: upright in vertical text flow using the text-orientation property or the text-combine-upright property.
+    if (((generalCategoryMask & U_GC_M_MASK) || (generalCategoryMask & U_GC_L_MASK)) && !isEastAsianFullWidth(character))
+        return CharacterClass::NonIdeographLetter;
+
+    if (isNonIdeographicNumeral(character, generalCategoryMask))
+        return CharacterClass::NonIdeographNumeral;
+
+    if (isFullWidthOpeningPunctuation(character, generalCategoryMask))
+        return CharacterClass::FullWidthOpeningPunctuation;
+
+    if (isFullWidthClosingPunctuation(character, generalCategoryMask))
+        return CharacterClass::FullWidthClosingPunctuation;
+
+    // FIXME: implement remaining classes for text-autospace: punctuation
+    return CharacterClass::Other;
+}
```

Comment:
The if-clauses in this function should be structured more efficiently.

- Since you have a category mask, use it more aggressively to filter things out. For example, you can check for [U_GC_P_MASK](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/uchar_8h.html#a84973d3e336d9858aac0bc31d39ffc5f) before doing more specific punctuation checks.
- Combine bitmasks together so you can check them simultaneously, e.g. (U_GC_M_MASK | U_GC_L_MASK)
- Exit early for things you know are not relevant, if you're already incorporating a conditional check for them. For example, once have checked for (U_GC_M_MASK | U_GC_L_MASK) we know that none of the rest of the function applies, so we can return instead of falling through to the punctuation checks. Because of short-circuiting, every clause in || or && is essentially an if statement -- if we know what to do for the else case, make that if-else explicit and make the short-circuit short-circuit the rest of the function.